core/types/data_source.go

```diff
@@ -221,20 +218,52 @@ func (s DataSourceDefinition) DeepClone() DataSourceDefinition {
 	if s.SourceType != nil {
 		switch dsn := s.SourceType.oneOfProto().(type) {
 		case *vegapb.DataSourceDefinition_External:
-			cpy = &DataSourceDefinition{
-				SourceType: &DataSourceDefinitionExternalx{
-					External: &DataSourceDefinitionExternal{
-						SourceType: &DataSourceSpecConfiguration{
-							Signers: SignersFromProto(dsn.External.GetOracle().Signers),
-							Filters: DataSourceSpecFiltersFromProto(dsn.External.GetOracle().Filters),
-						},
-					},
-				},
+			if dsn.External != nil {
```

Comment:
This looks quite repetitious, you have a similar type switch and nil checks for things like internal time a bit further down. Rather than having this `oneOfProto().(type)` switch wrapped around a `dsn.External.SourceTyp.(type)` switch, wouldn't it be possible to have an enum type of sorts, with some getters where you can do something like this:

```go
    switch s.SourceType {
    case SourceTypeExternal:
        if oracle := s.SourceType.oneOfProto().GetExternalOracle(); oracle != nil {
            signers := []*datapb.Signer{}
            // handle here
            return
        }
    case SourceTypeInternal:
        if tm := s.SourceType.oneOfProto().InternalTime(); tm != nil {
        }
    }
```

There still will be type assertions required in the actual getters, but this way, there would be less need to perform all the nil checks inside this many levels (we heve the outer `s.SourceType != nil`, a type switch on the `oneOfProto()`, with 2 nil checks, with another type switch. That's quite a lot of levels

Reply:
This has been refactored everywhere, thanks!